src/libclipper/include/clipper/selection_policies.hpp

```diff
@@ -0,0 +1,287 @@
+    return Derived::initialize(candidate_models);
+  };
+
+  static State add_models(State state,
+                          const std::vector<VersionedModelId>& new_models) {
+    return Derived::add_models(state, new_models);
+  };
+
+  // Used to identify a unique selection policy instance. For example,
+  // if using a bandit-algorithm that does not tolerate variable-armed
+  // bandits, one could hash the candidate models to identify
+  // which policy instance corresponds to this exact set of arms.
+  // Similarly, it provides flexibility in how to deal with different
+  // versions of the same arm (different versions of same model).
+  static long hash_models(
+      const std::vector<VersionedModelId>& candidate_models) {
+    return Derived::hash_models(candidate_models);
+  }
+
+  // Query Pre-processing: select models and generate tasks
+  static std::vector<PredictTask> select_predict_tasks(State state, Query query,
+                                                       long query_id) {
+    return Derived::select_predict_tasks(state, query, query_id);
+  }
+
+  // TODO: change this method name
+  // TODO: I think it may make sense to decouple combine_predictions()
+  // from select_predict_tasks in some cases
+  static Output combine_predictions(State state, Query query,
+                                    std::vector<Output> predictions) {
+    return Derived::combine_predictions(std::forward(state), query,
+                                        predictions);
+  }
+
+  /// When feedback is received, the selection policy can choose
+  /// to schedule both feedback and prediction tasks. Prediction tasks
+  /// can be used to get y_hat for e.g. updating a bandit algorithm,
+  /// while feedback tasks can be used to optionally propogate feedback
+  /// into the model containers.
+  static std::pair<std::vector<PredictTask>, std::vector<FeedbackTask>>
+  select_feedback_tasks(State state, FeedbackQuery query, long query_id) {
+    return Derived::select_feedback_tasks(std::forward(state), query, query_id);
+  }
+
+  /// This method will be called if at least one PredictTask
+  /// was scheduled for this piece of feedback. This method
+  /// is guaranteed to be called sometime after all the predict
+  /// tasks scheduled by `select_feedback_tasks` complete.
+  static State process_feedback(State state, Feedback feedback,
+                                std::vector<Output> predictions) {
+    return Derived::process_feedback(std::forward(state), feedback,
+                                     predictions);
+  }
+
+  static std::string serialize_state(State state) {
+    return Derived::serialize_state(state);
+  }
+
+  static State deserialize_state(const std::string& bytes) {
+    return Derived::deserialize_state(bytes);
+  }
+
+  /**
+   * Human readable debug string for the state
+   */
+  static std::string state_debug_string(const State& state) {
+    return Derived::state_debug_string(state);
```

Comment:
Can you implement `state_debug_string` for each implementation? The string should be a human-readable representation of the state.